### ISSUE TITLE
SCM GIT job exports different on different servers

### ISSUE THREAD

**Issue by matthewsrogers** [rundeck/rundeck#2483]
Issue type: Bug report/Enhancement Request

**My Rundeck detail**

* Rundeck version: 2.7.3-1 (same symptoms on 2.8.x as well)
* install type: (rpm,deb,war,launcher?) RPM
* OS Name/version: CentOS 6
* DB Type/version: mysql? postgres? h2?.. H2

Several members of my team and I each have our own development VM, each running Rundeck 2.7.3-1.  We have a single corporate GitHub server that hosts our Rundeck job repository.  I set up the SCM Git Import plugin on each VM, and I'm able to import each job from the repo into each Rundeck instance just fine.  As soon as I set up the export plugin, however, nearly every job is marked as "Export Status: Modified, Import Status: Up to date".  When I look at the diff, the only changes are rearrangements of the XML blocks, no real changes.  Every server behaves a bit differently as far as the order it uses to export the job into XML.  This is a problem because any time anybody makes a change to one job, many other jobs are changed for no reason.

**Expected Behavior**
Export behavior should be identical across all Rundeck servers.  Lines/blocks of XML should be exported in the same order no matter the server.  When someone submits a pull request, it should only contain changes they made.  It should NOT contain changes to other jobs that were modified by Rundeck for no reason.

**Actual Behavior**
Export behavior varies across servers so even just setting up the export plugin makes changes to nearly every job.  If multiple dev servers are set up to create/modify jobs, each will make different spurious changes from the other.  Sure, nothing functional in the job changes, but it sure makes a mess of GIT.

**How to reproduce Behavior**
Set up 2 servers with the same version of Rundeck.  Configure the import plugin and import the jobs from a repo.  Set up the export plugin, and you should instantly see changes to the jobs you just imported.

**Comment by gschueler** [comment]
thanks, is there a consistent block that is not showing up correctly?

is it possible for you to show the buggy diff of your job?

**Comment by matthewsrogers** [comment]
Absolutely.  I'll post a couple (redacted as necessary).

This first one is the most prevalent and most annoying:
```@@ -24,7 +24,7 @@
       </command>
       <pluginConfig>
         <WorkflowStrategy>
-          <node-first />
+          <node-first></node-first>
         </WorkflowStrategy>
       </pluginConfig>
     </sequence>
```

This following example of how just setting up the SCM GIT export plugin modifies job definitions by rearranging blocks of XML (no functional changes).  So even though the job definitions are the same functionally, GIT sees the rearrangement of lines as legitimate changes:
```
@@ -37,6 +37,15 @@
     <nodeFilterEditable>false</nodeFilterEditable>
     <notification>
       <onfailure>
+        <plugin type='SlackNotification'>
+          <configuration>
+            <entry key='room' value='#redacted' />
+            <entry key='icon_url' value='' />
+            <entry key='webhookUrl' value='https://hooks.slack.com/services/redacted' />
+            <entry key='external_template' value='' />
+            <entry key='username' value='' />
+          </configuration>
+        </plugin>
         <plugin type='NRReinject'>
           <configuration>
             <entry key='nrhost' value='Production' />
@@ -53,26 +62,8 @@
             <entry key='username' value='redacted' />
           </configuration>
         </plugin>
-        <plugin type='SlackNotification'>
-          <configuration>
-            <entry key='room' value='#redacted' />
-            <entry key='icon_url' value='' />
-            <entry key='webhookUrl' value='https://hooks.slack.com/services/redacted' />
-            <entry key='external_template' value='' />
-            <entry key='username' value='' />
-          </configuration>
-        </plugin>
       </onfailure>
       <onstart>
-        <plugin type='SlackNotification'>
-          <configuration>
-            <entry key='room' value='#redacted' />
-            <entry key='external_template' value='' />
-            <entry key='webhookUrl' value='https://hooks.slack.com/services/redacted' />
-            <entry key='icon_url' value='' />
-            <entry key='username' value='' />
-          </configuration>
-        </plugin>
         <plugin type='MysqlLogNotification'>
           <configuration>
             <entry key='dbhost' value='redacted' />
@@ -82,6 +73,15 @@
             <entry key='job_type' value='Remediation' />
           </configuration>
         </plugin>
+        <plugin type='SlackNotification'>
+          <configuration>
+            <entry key='room' value='#redacted' />
+            <entry key='external_template' value='' />
+            <entry key='webhookUrl' value='https://hooks.slack.com/services/redacted' />
+            <entry key='icon_url' value='' />
+            <entry key='username' value='' />
+          </configuration>
+        </plugin>
       </onstart>
       <onsuccess>
         <plugin type='MysqlLogNotification'>
```
I proved there were no changes by sorting a job definition file from Rundeck and the corresponding job definition in my GIT repo, by running: diff <(sort file_from_rundeck) <(sort file_from_repo).  Diff showed no differences at all.

Please let me know if I can provide any more information.  I've written a work-around script to handle SCM export functionality (detect "real" changes and copy them to the repo directory).  It's not pretty, but it works.  Being able to use the UI would make things much easier.

Thanks!

**Comment by gschueler** [comment]
@matthewsrogers thanks, this is helpful

**Comment by matthewsrogers** [comment]
Hi @gschueler, I've been anxiously looking forward to this bug fix.  I noticed that you have a 2.9 beta release available for download.  You have this as a 2.8.3 milestone.  Unfortunately, I can't see any indication of this fix in the change logs for the 2.9 beta release.  Do you have an ETA on when this bug fix will be included in a release?

Thanks!

**Comment by gschueler** [comment]
@matthewsrogers we have put a release date on [2.8.3](https://github.com/rundeck/rundeck/milestone/55) for this friday (6/30).

the change will be included in that release

it is also part of the current 2.9.0-BETA1 release if you want to try it
